rllib/models/specs/specs_base.py

```diff
@@ -0,0 +1,209 @@
+import abc
+from typing import Any, Optional, Dict, List, Tuple, Union
+
+from ray.rllib.utils.annotations import DeveloperAPI
+from ray.rllib.utils.typing import TensorType
+
+_INVALID_INPUT_DUP_DIM = "Duplicate dimension names in shape ({})"
+_INVALID_INPUT_Unknown_DIM = "Unknown dimension name {} in shape ({})"
+_INVALID_INPUT_NON_POSITIVE = "Dimension {} in ({}) has non-positive size"
+_INVALID_INPUT_NONINT_DIM = "Dimension {} in ({}) cannot be non-integer, got {}"
+_INVALID_SHAPE = "Expected shape {} but found {}"
+_INVALID_DTYPE = "Expected dtype {} but found {}"
+
+
+@DeveloperAPI
+class TensorSpecs(abc.ABC):
+    """A base class that specifies the shape and dtype of a tensor.
+
+    Args:
+        shape: A string representing einops notation of the shape of the tensor.
+            For example, "B C" represents a tensor with two dimensions, the first
+            of which has size B and the second of which has size C. shape should
+            consist of unique dimension names. For example having "B B" is invalid.
+        dtype: The dtype of the tensor. If None, the dtype is not checked during
+            validation. Also during Sampling the dtype is set the default dtype of
+            the backend.
+        shape_vals: An optional dictionary mapping some dimension names to their
+            values. For example, if shape is "B C" and shape_vals is {"C": 3}, then
+            the shape of the tensor is (B, 3). B is to be determined during
+            run-time but C is fixed to 3.
+
+    Examples:
+        >>> spec = TensorSpec("b h", h=128, dtype=tf.float32)
+        >>> spec.shape  # ('b', 128)
+        >>> spec.validate(torch.rand(32, 128, dtype=torch.float64))  # passes
+        >>> spec.validate(torch.rand(32, 64, dtype=torch.float64))   # raises ValueError
+        >>> spec.validate(torch.rand(32, 128, dtype=torch.float64))  # raises ValueError
```

Comment:
isn't this the same as like 35? needs updated example

Reply:
had a typo, fixed.

Reply:
String format is used in `np.einsum/torch.einsum` which is a bit more readable/writeable IMO than something like `('b', 't', 'a', 'h')`